### PR TITLE
Decided message signers fix

### PR DESCRIPTION
**Pull request by amir-blox**
Resolve https://github.com/bloxapp/ssv/issues/430

IBFT messages were purged once the instance decided or stopped, the behavior changed in this PR:
- when instance finished w/ decided - ctrl starts to listen for late commit messages with timeout and purge messages once done
- when instance finished w/o decided - ctrl purges messages

### CODE REVIEW

File: ibft/controller/controller_decided.go [hunk 1]
```diff
@@ -119,6 +129,22 @@ func (i *Controller) decidedMsgKnown(msg *proto.SignedMessage) (bool, error) {
 	return found, nil
 }
 
+// checkDecidedMessageSigners checks if signers of existing decided includes all signers of the newer message
+func (i *Controller) checkDecidedMessageSigners(msg *proto.SignedMessage) (bool, error) {
```
Comment (alonmuroch) [inline ibft/controller/controller_decided.go:133]: @amir-blox tests for those conditions?

File: ibft/controller/controller_running_instance.go [hunk 2]
```diff
@@ -107,33 +122,41 @@ func (i *Controller) instanceStageChange(stage proto.RoundState) (bool, error) {
...
 			if netMsg := i.msgQueue.PopMessage(idxKey); netMsg != nil && netMsg.SignedMessage != nil {
 				if netMsg.SignedMessage.Message == nil || netMsg.SignedMessage.Message.Type != proto.RoundState_Commit {
 					// not a commit message -> skip
 					continue
 				}
 				logger := i.logger.With(zap.Uint64("seq", netMsg.SignedMessage.Message.SeqNumber),
-					zap.String("identifier", string(netMsg.SignedMessage.Message.Lambda)))
-				if err := runningInstance.CommitMsgValidationPipeline().Run(netMsg.SignedMessage); err != nil {
+					zap.Uint64s("signers", netMsg.SignedMessage.SignerIds))
+				if err := instance.CommitMsgValidationPipelineV0(identifier, seq, i.ValidatorShare).Run(netMsg.SignedMessage); err != nil {
```
Comment (alonmuroch) [inline ibft/controller/controller_running_instance.go:144]: @amir-blox better use the fork and not the implementation directly

File: ibft/controller/controller_running_instance.go [hunk 3]
```diff
@@ -107,33 +122,41 @@ func (i *Controller) instanceStageChange(stage proto.RoundState) (bool, error) {
...
+
 	f := func(stopper tasks.Stopper) (interface{}, error) {
 	loop:
 		for {
 			if stopper.IsStopped() {
 				break loop
 			}
-			idxKey := msgqueue.IBFTMessageIndexKey(runningInstance.State().Lambda.Get(), runningInstance.State().SeqNumber.Get())
 			if netMsg := i.msgQueue.PopMessage(idxKey); netMsg != nil && netMsg.SignedMessage != nil {
 				if netMsg.SignedMessage.Message == nil || netMsg.SignedMessage.Message.Type != proto.RoundState_Commit {
```
Comment (alonmuroch) [inline ibft/controller/controller_running_instance.go:138]: why check msg nil? pipeline should take care of it no?

File: ibft/instance/commit.go [hunk 4]
```diff
@@ -65,12 +76,17 @@ func (i *Instance) CommitMsgValidationPipeline() pipeline.Pipeline {
 
 // CommitMsgValidationPipelineV0 is version 0
 func (i *Instance) CommitMsgValidationPipelineV0() pipeline.Pipeline {
+	return CommitMsgValidationPipelineV0(i.State().Lambda.Get(), i.State().SeqNumber.Get(), i.ValidatorShare)
+}
+
+// CommitMsgValidationPipelineV0 is version 0 of commit message validation
+func CommitMsgValidationPipelineV0(identifier []byte, seq uint64, share *storage.Share) pipeline.Pipeline {
```
Comment (alonmuroch) [inline ibft/instance/commit.go:83]: that seems redundant ..

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: ibft/instance/commit.go (hunk 4 above)
Comment (amir-blox) [reply ibft/instance/commit.go:83]: its a function so the exporter could use it as well

Comment (alonmuroch) [reply ibft/instance/commit.go:83]: we have the instance which has that function why create another?

Comment (amir-blox) [reply ibft/instance/commit.go:83]: late commit is done on controller level when the instance finished and on exporter, in both cases we don't have an instance..

**Comment by amir-blox** [comment]
The ibft instance is not available in late commit scenarios:
1. ibft controller after the instance was cleared
2. exporter

Therefore in this PR, a static commit message validation function was added to overcome the missing instance object.
It should be refactored in the future to be accessed via the fork

**Pull request merged**